p4_symbolic/symbolic/action.cc

```diff
@@ -326,7 +326,8 @@ absl::Status EvaluateAction(const ir::Action &action,
         parameters.at(i);
     const std::string &parameter_name = parameter.param().name();
     ASSIGN_OR_RETURN(z3::expr parameter_value,
-                     util::IrValueToZ3Expr(args.at(i - 1).value()));
+                     environment->value_formatter.FormatP4RTValue(
+                         parameter_name, args.at(i - 1).value()));
```

Comment:
Parameter name is the wrong thing there, I think it should be parameter type (vrf_t), right? Like, VRFs in an L3 match should be translated consistently with the action that sets it in the VRF classifier table?

Reply:
This is a very important point, we cannot do that currently unfortunately, primarily because we do not have access to the annotation on the type, it is part of p4info, and I could not find it exposed by p4-pdpi, but most important becasue the scenario that you are describing would not use the name "vrf_t" in the bmv2 json (and thus in the action implementation for example), because the bmv2 just inlines the type def directly.

In my docs, I discuss this at length, the current approach will work for the sai program, I discuss why there, but may not work for other programs if they are too liberal in how they use their string values because of what you are hinting at.

Do you have some time to discuss this today?

Reply:
I may not have all of the context, but PDPI should give you all the info you need. A parameter that is string encoded would look like this:

```
        params_by_name {
          key: "str"
          value {
            param {
              id: 5
              name: "str"
              type_name {
                name: "string_id_t"
              }
            }
            format: STRING
          }
        }
```

You know that this has the type string_id_t, and you know it has the @p4rt_translation(.., string) annotation, because format == Format::STRING.

Is that not enough?

Reply:
I am using this now, it is enough for this. I phrased what I am saying badly.

Everything related to table entries or p4info is sufficiently handled by pdpi. The problem is with things that do not appear in p4info and are only defined in bmv2 JSON (e.g. a header field). All type information about these things is lost because bmv2 strips it out.

Luckily, we can do something that works most of the time. A field that is translated is probably gonna be used as some table match key somewhere (otherwise why is it used), if thats the case, we can find out the type information from p4info/pdpi during the match key evaluation, and store it, so we are now aware of the type of the field. Not 100% fool proof but works well for most cases (and for the SAI program).

That's how the current implementation works. I updated the docs as well.